[Concurrency] Implement type checking for 'async let' declarations.

DougGregor opened the pull request:
Implement parsing and type checking for 'async let' declarations. 'async let' allows one to create child tasks that execute concurrently with the current task. This covers:
* Declaring `async let` declarations within asynchronous functions
* Ensuring that access to an `async let` declaration is covered by `await` (and `try`, if the initializer could throw an error)
* Treating the initializer as a concurrently-executing child task, by injecting it into an escaping autoclosure, which (e.g.) takes it out of the actor isolation domain

Covers the type checking portion of rdar://70101851.

DougGregor committed:
[Concurrency] Parse 'async let' declarations.

Perform very basic semantic analysis for their well-formedness.

DougGregor committed:
[Concurrency] Require references to 'async let' to have an 'await'.

Extend effects checking to ensure that each reference to a variable
bound by an 'async let' is covered by an "await" expression and occurs
in a suitable context.

DougGregor committed:
[Concurrency] Diagnose 'async let' declarations in non-async contexts.

ktoso commented on test/decl/var/async_let.swift:
```diff
@@ -0,0 +1,25 @@
+  _ = z2
+  _ = x2
+  x = 1
+}
```
I guess you just didn't get to it yet :) 

async let being wrong inside of a sync func:

```swift
func testSyncFunc() {
   async let x2 = 1 // expected-error
}
```

DougGregor replied:
See the third commit I just pushed ;)

ktoso replied:
ha, excellent :)

rintaro commented:
Could you add test cases for multiple patterns? like
```swift
async let veggies = try chopVegetables(), meat = marinateMeat()
```

DougGregor committed:
[Concurrency] Wrap the initializer of 'async let' in an autoclosure call.

The initializer of an 'async let' is executed as a separate child task
that will run concurrently with the main body of the function. Model
the semantics of this operation by wrapping the initializer in an
async, escaping autoclosure (representing the evaluation of the child
task), and then a call to that autoclosure (to

This is useful both for actor isolation checking, which needs to treat
the initializer as executing in concurrent code, and also (eventually)
for code generation, which needs to have that code in a closure so
that it can be passed off to the task-creation functions.

There are a number of issues with this implementation producing
extraneous diagnostics due to this closure transformation, which will
be addressed in a follow-up commit.

DougGregor committed:
[Concurrency] Don't diagnose missing "self." in async let autoclosures.

DougGregor committed:
[Concurrency] Improve await/try checking in 'async let' initializers.

Customize diagnostics when an 'await' is missing in an 'async let'
initializer. While here, fix the coverage checking so we also diagnose
a missing 'try'.

DougGregor committed:
[Concurrency] Require 'try' on accessing 'async let' with throwing initializer.

When an 'async let' initializer can throw, any access to one of the
variables in the 'async let' can also throw, so require such accesses
to be annotated with 'try'.

DougGregor committed:
[Concurrency] Add an async let example with multiple initializer clauses

DougGregor commented:
> Could you add test cases for multiple patterns? like
> 
> ```swift
> async let veggies = try chopVegetables(), meat = marinateMeat()
> ```

Sure, I've added it, plus a bunch of other tests.

DougGregor commented:
@swift-ci please smoke test

ktoso reviewed (approved):
Looking great, just some small inline comments 👍

DougGregor commented on include/swift/AST/Decl.h:
```diff
@@ -1784,6 +1784,9 @@ class PatternBindingDecl final : public Decl,
   /// Is the pattern binding entry for this variable  currently being computed?
   bool isComputingPatternBindingEntry(const VarDecl *vd) const;
 
+  /// Is this an "async let" declaration.
```
Sure, I'll fix in a follow-up PR.

DougGregor commented on test/Concurrency/async_let_isolation.swift:
```diff
@@ -0,0 +1,29 @@
+
+  func testAsyncLetIsolation() async {
+    async let x = self.synchronous()
+    // expected-error @-1{{actor-isolated instance method 'synchronous()' is unsafe to reference in code that may execute concurrently}}
```
I agree that we could do better with the "why?" here, e.g., we could call out that we're in a child task.

DougGregor commented on test/expr/unary/async_await.swift:
```diff
@@ -144,3 +144,53 @@ func invalidAsyncFunction() async {
+  _ = await try x3
+  _ = await x4
+  _ = await x5
+}
```
The way we've structured effects checking in the type checker prevents us from doing flow-sensitive analysis like this.

DougGregor replied:
The paragraph you cite isn't in the proposal because of this ;)

ktoso replied:
Hah got it :) Thanks

DougGregor commented on test/decl/var/async_let.swift:
```diff
@@ -0,0 +1,37 @@
+  _ = await z2
+  _ = await x2
+  await x = 1
+  _ = y2
```
Due to the semantic error in the declaration of `y2`, we don't end up checking for `await` here.

ktoso replied:
Got it thanks

DougGregor commented on test/decl/var/async_let.swift:
```diff
@@ -0,0 +1,37 @@
+  _ = await z1
+  _ = await z2
+  _ = await x2
+  await x = 1
```
The `async var` is already an error, so there is no need to diagnose this. An assignment expression is fine to put inside an `await`, and there are no assignable `async let`s.

ktoso replied:
okey so this is an assignment inside an await... Oh that's weird... I guess this is in order to allow `await x = y` where y is async...? I guess a bit unexpected but not wrong I guess...

await x = 1 I would expect to at least note that it's an unnecessary await? 🤔

DougGregor committed:
[Concurrency] Fix a comment.

kastiglione commented on test/decl/var/async_let.swift:
```diff
@@ -0,0 +1,37 @@
+
+// REQUIRES: concurrency
+
+async let x = 1 // okay
```
Curious: why allow non-async initializers?

DougGregor replied:
The work you want to run concurrently could be synchronous work. Why not?

kastiglione replied:
I didn't think that one through, sorry. I assumed (I don't know why) that concurrent code would be via async functions.

kastiglione commented on test/expr/unary/async_await.swift:
```diff
@@ -144,3 +144,53 @@ func invalidAsyncFunction() async {
+}
+
+func testAsyncLet() async throws {
+  async let x = await getInt()
+  print(x) // expected-error{{reference to async let 'x' is not marked with 'await'}}
+  print(await x)
```
Compared to the proposals and other docs, this is the first I've seen `await` being used on an `async let` initializer. Is the `await` keyword going to be required for initializers that call async functions? Are there two suspension points, at the initializer and at the subsequent read (on line 161)?

I would think the `await` would only be required in a case like this:

```
async let y = getY()
async let x = await calculateX(y)
```

DougGregor replied:
I (intentionally) deviated from the proposal document, because as I was writing the tests I felt like skipping the `await` was arbitrary and confusing.

kastiglione replied:
oh, why arbitrary to not have it? My feeling was the opposite, I thought that adding the `await` reads like a suspension point, but if it's not a suspension point then the `await` seems arbitrary. And as I mentioned, the two `await`s also struck me as potentially confusing. I'm not arguing a position, just wanted to understand this use, and to share my perspective, for what it's worth.

DougGregor replied:
We should take this over to the forums. You're right that the `async let` itself isn't necessarily a suspension point, so in that sense, the `await` might be misleading. On the other hand, it tells you whether you have asynchronous code within the initializer, which... might?... be interesting.
